lib/colibri2_mappings.ml

```diff
@@ -0,0 +1,989 @@
+           in
+           let run_1 env subst = mk_eq env subst xt_str term_str in
+           let run_2 env subst = mk_eq env subst xt term in
+           Colibri2_theories_quantifiers.InvertedPath.add_callback env pattern_1
+             run_1;
+           Colibri2_theories_quantifiers.InvertedPath.add_callback env pattern_2
+             run_2
+       in
+       add_string_axiom env int_to_string string_to_int DTy.int;
+       add_string_axiom env real_to_string string_to_real DTy.real;
+       add_string_axiom env f32_to_string string_to_f32 float32_ty;
+       add_string_axiom env f64_to_string string_to_f64 float64_ty *)
+
+    let encode_expr e = encore_expr_aux e
+    let pp_smt ?status:_ _ _ = ()
+    let interrupt () = ()
+
+    module Solver = struct
+      let mk_scheduler () =
+        let scheduler = Scheduler.new_solver ~learning:false () in
+        Scheduler.init_theories
+          ~theories:
+            ( Colibri2_theories_bool.Ite.th_register :: LRA.LRA.th_register
+            :: Colibri2_theories_fp.Fp.th_register
+            :: Colibri2_core.ForSchedulers.default_theories () )
+          scheduler;
+        scheduler
+
+      let make ?logic:_ () =
+        let scheduler = mk_scheduler () in
+        let ctx = Scheduler.get_context scheduler in
+        { scheduler
+        ; pushpop = []
+        ; state = `Search
+        ; status_colibri = Context.Ref.create ctx `No
+        ; decls = DTerm.Const.S.empty
+        }
+
+      let add_simplifier s = s
+
+      let clone { pushpop; state; status_colibri; decls; _ } =
+        let scheduler = mk_scheduler () in
+        { scheduler; pushpop; state; status_colibri; decls }
+
+      let push st = st.pushpop <- Scheduler.push st.scheduler :: st.pushpop
+
+      let rec pop st i =
+        assert (0 <= i);
+        match (i, st.pushpop) with
+        | 0, _ -> ()
+        | _, [] -> assert false
+        | 1, bp :: l ->
+          st.pushpop <- l;
+          Scheduler.pop_to st.scheduler bp
+        | n, _ :: l ->
+          st.pushpop <- l;
+          pop st (n - 1)
+
+      let reset s =
+        let scheduler = mk_scheduler () in
+        let ctx = Scheduler.get_context scheduler in
+        s.scheduler <- scheduler;
+        s.pushpop <- [];
+        s.state <- `Search;
+        s.status_colibri <- Context.Ref.create ctx `No;
+        s.decls <- DTerm.Const.S.empty
+
+      let add s es =
+        Scheduler.add_assertion s.scheduler (fun d ->
+            let es' =
+              List.map
+                (encore_expr_aux ~record_sym:(fun c ->
+                     s.decls <- DTerm.Const.S.add c s.decls ) )
+                es
+            in
+            List.iter
+              (fun e ->
+                let n = Colibri2_core.Ground.convert d e in
+                Colibri2_core.Egraph.register d n;
+                Colibri2_theories_bool.Boolean.set_true d n )
+              es' )
+
+      let check s es =
+        add s es;
+        Scheduler.check_sat s.scheduler
```

Comment:
The way I encoded this using Z3 was for the expressions `es` to not be added to the solver. Because, IIRC Z3's `check` actually behaves like a `check-sat-assuming` with the argument `es` being the assumptions.

Do you think this will be problematic? We should probably add a label to `es` something like: `Val check : solver -> assumptions:expr list`